src/Shared/CoreCLRAssemblyLoader.cs

```diff
@@ -108,33 +106,16 @@ private Assembly LoadUsingPluginContext(string fullPath)
             }
         }
 
-        private Assembly TryGetWellKnownAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
-        {
-            if (!MSBuildLoadContext.WellKnownAssemblyNames.Contains(assemblyName.Name))
-            {
-                return null;
-            }
-
-            // Ensure we are attempting to load a matching version
-            // of the Microsoft.Build.* assembly.
-            assemblyName.Version = _currentAssemblyVersion;
-
-            string[] searchPaths = [Assembly.GetExecutingAssembly().Location];
-            return TryResolveAssemblyFromPaths(context, assemblyName, searchPaths);
-        }
-
         private Assembly TryResolveAssembly(AssemblyLoadContext context, AssemblyName assemblyName)
         {
             lock (_guard)
             {
-                Assembly assembly = TryGetWellKnownAssembly(context, assemblyName);
-
-                if (assembly != null)
+                if (MSBuildLoadContext.WellKnownAssemblyNames.Contains(assemblyName.Name))
                 {
-                    return assembly;
+                    return MSBuildLoadContext.ThisAssemblyLoadContext.LoadFromAssemblyName(assemblyName);
```

Comment:
I think this achieves the same goal in a much more straightforward way. And since we load a specific name in a specific context we don't need to patch the version.